File: engine/tilemap.py
"""
Tile-based map system.

Maps are stored as JSON and rendered in the viewport as a 2D overhead view.
Each tile is 20×20 pixels in the viewport (configurable).
"""
from __future__ import annotations
import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple


# ---------------------------------------------------------------------------
# Tile definitions
# ---------------------------------------------------------------------------

TILE_DEFS: Dict[int, Dict[str, Any]] = {
    0:  {"id": 0, "name": "floor",       "passable": True,  "glyph": ".",  "art": "FLOOR"},
    1:  {"id": 1, "name": "wall",        "passable": False, "glyph": "#",  "art": "WALL"},
    2:  {"id": 2, "name": "door_closed", "passable": True,  "glyph": "+",  "art": "DOOR"},
    3:  {"id": 3, "name": "door_open",   "passable": True,  "glyph": "/",  "art": "FLOOR"},
    4:  {"id": 4, "name": "stairs_down", "passable": True,  "glyph": ">",  "art": "STAIRS"},
    5:  {"id": 5, "name": "stairs_up",   "passable": True,  "glyph": "<",  "art": "STAIRS"},
    6:  {"id": 6, "name": "water",       "passable": False, "glyph": "~",  "art": "WATER"},
    7:  {"id": 7, "name": "chest",       "passable": True,  "glyph": "$",  "art": "CHEST"},
    8:  {"id": 8, "name": "altar",       "passable": True,  "glyph": "&",  "art": "SPECIAL"},
    9:  {"id": 9, "name": "pillar",      "passable": False, "glyph": "O",  "art": "WALL"},
    10: {"id":10, "name": "void",        "passable": False, "glyph": " ",  "art": "VOID"},
    11: {"id":11, "name": "fireplace",   "passable": False, "glyph": "*",  "art": "SPECIAL"},
}

WALL    = 1
FLOOR   = 0
DOOR    = 2
STAIRS_D = 4
STAIRS_U = 5
WATER   = 6


# ---------------------------------------------------------------------------
# Map entity
# ---------------------------------------------------------------------------

@dataclass
class MapEntity:
    entity_type: str       # "npc", "item", "player_start", "trigger"
    entity_id:   str
    x:           int
    y:           int
    properties:  Dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Tilemap
# ---------------------------------------------------------------------------

@dataclass
class TileMap:
    """
    A 2D grid of tile IDs with entity placements.
    """
    map_id:     str
    name:       str
    width:      int
    height:     int
    tiles:      List[List[int]]              # tiles[y][x]
    entities:   List[MapEntity]             = field(default_factory=list)
    connections: Dict[str, Optional[str]]   = field(default_factory=dict)
    scene_id:   Optional[str]              = None
    metadata:   Dict[str, Any]             = field(default_factory=dict)

    # Runtime state
    revealed:   List[List[bool]]           = field(default_factory=list)  # fog of war

    def __post_init__(self) -> None:
        if not self.revealed:
            self.revealed = [[False] * self.width for _ in range(self.height)]

    # -----------------------------------------------------------------------
    # Tile access
    # -----------------------------------------------------------------------

    def get(self, x: int, y: int) -> int:
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.tiles[y][x]
        return WALL  # out-of-bounds is a wall

    def set(self, x: int, y: int, tile_id: int) -> None:
        if 0 <= x < self.width and 0 <= y < self.height:
            self.tiles[y][x] = tile_id

    def is_passable(self, x: int, y: int) -> bool:
        td = TILE_DEFS.get(self.get(x, y), {})
        if not td.get("passable", False):
            return False
        for e in self.entities:
            if e.x == x and e.y == y and e.entity_type == "npc":
                return False
        return True

    def tile_name(self, x: int, y: int) -> str:
        return TILE_DEFS.get(self.get(x, y), {}).get("name", "unknown")

    def tile_glyph(self, x: int, y: int) -> str:
        return TILE_DEFS.get(self.get(x, y), {}).get("glyph", "?")

    # -----------------------------------------------------------------------
    # Fog of war
    # -----------------------------------------------------------------------

    def reveal_radius(self, cx: int, cy: int, radius: int = 5) -> None:
        """Mark tiles within radius as revealed."""
        for dy in range(-radius, radius + 1):
            for dx in range(-radius, radius + 1):
                if dx*dx + dy*dy <= radius*radius:
                    nx, ny = cx + dx, cy + dy
                    if 0 <= nx < self.width and 0 <= ny < self.height:
                        self.revealed[ny][nx] = True

    def is_revealed(self, x: int, y: int) -> bool:
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.revealed[y][x]
        return False

    # -----------------------------------------------------------------------
    # Entity helpers
    # -----------------------------------------------------------------------

    def entities_at(self, x: int, y: int) -> List[MapEntity]:
        return [e for e in self.entities if e.x == x and e.y == y]

    def find_entity(self, entity_id: str) -> Optional[MapEntity]:
        for e in self.entities:
            if e.entity_id == entity_id:
                return e
        return None

    def remove_entity(self, entity_id: str) -> bool:
        for e in list(self.entities):
            if e.entity_id == entity_id:
                self.entities.remove(e)
                return True
        return False

    def player_start(self) -> Tuple[int, int]:
        for e in self.entities:
            if e.entity_type == "player_start":
                return e.x, e.y
        return self.width // 2, self.height // 2

    # -----------------------------------------------------------------------
    # Serialization
    # -----------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "id": self.map_id,
            "name": self.name,
            "width": self.width,
            "height": self.height,
            "tiles": self.tiles,
            "entities": [
                {
                    "type": e.entity_type,
                    "id": e.entity_id,
                    "x": e.x,
                    "y": e.y,
                    "properties": e.properties,
                }
                for e in self.entities
            ],
            "connections": self.connections,
            "scene_id": self.scene_id,
            "metadata": self.metadata,
        }

    def save(self, path: str) -> None:
        with open(path, "w") as f:
            json.dump(self.to_dict(), f, indent=2)

    @classmethod
    def from_dict(cls, d: dict) -> "TileMap":
        entities = [
            MapEntity(
                entity_type=e["type"],
                entity_id=e["id"],
                x=e["x"],
                y=e["y"],
                properties=e.get("properties", {}),
            )
            for e in d.get("entities", [])
        ]
        return cls(
            map_id=d.get("id", "map"),
            name=d.get("name", "Unnamed"),
            width=d["width"],
            height=d["height"],
            tiles=d["tiles"],
            entities=entities,
            connections=d.get("connections", {}),
            scene_id=d.get("scene_id"),
            metadata=d.get("metadata", {}),
        )

    @classmethod
    def load(cls, path: str) -> "TileMap":
        with open(path) as f:
            return cls.from_dict(json.load(f))

    @classmethod
    def blank(cls, width: int = 20, height: int = 15, map_id: str = "new_map") -> "TileMap":
        """Create an empty map filled with walls, with a floor room inside."""
        tiles = [[WALL] * width for _ in range(height)]
        for y in range(1, height - 1):
            for x in range(1, width - 1):
                tiles[y][x] = FLOOR
        return cls(
            map_id=map_id,
            name="New Map",
            width=width,
            height=height,
            tiles=tiles,
            entities=[MapEntity("player_start", "player_start", width//2, height//2)],
        )

    @classmethod
    def from_string(cls, s: str, map_id: str = "map") -> "TileMap":
        """
        Parse a map from an ASCII string.
        '#'=wall  '.'=floor  '+'=door  '>'=stairs_down  '<'=stairs_up
        '~'=water  '@'=player start
        """
        CHAR_TO_TILE = {
            '#': WALL, '.': FLOOR, '+': DOOR, '>': STAIRS_D,
            '<': STAIRS_U, '~': WATER, ' ': 10, '@': FLOOR,
            '*': 11, '&': 8, '$': 7, 'O': 9,
        }
        lines = s.strip("\n").splitlines()
        height = len(lines)
        width  = max(len(l) for l in lines)
        tiles  = []
        entities = []
        for y, line in enumerate(lines):
            row = []
            for x, ch in enumerate(line.ljust(width)):
                row.append(CHAR_TO_TILE.get(ch, WALL))
                if ch == '@':
                    entities.append(MapEntity("player_start", "player_start", x, y))
            tiles.append(row)
        return cls(
            map_id=map_id, name=map_id,
            width=width, height=height,
            tiles=tiles, entities=entities,
        )

File: js/engine/tilemap.js
// tilemap.js – Tile map system with fog-of-war
'use strict';
var DCE = window.DCE || {};
window.DCE = DCE;

DCE.TILE = { FLOOR:0, WALL:1, DOOR:2, DOOR_OPEN:3, STAIRS_D:4, STAIRS_U:5,
             WATER:6, CHEST:7, ALTAR:8, PILLAR:9, VOID:10, FIRE:11 };

DCE.TILE_DEFS = {
  0:  { name:'Floor',     glyph:'·', passable:true,  art:'FLOOR'  },
  1:  { name:'Wall',      glyph:'█', passable:false, art:'WALL'   },
  2:  { name:'Door',      glyph:'+', passable:true,  art:'DOOR'   },
  3:  { name:'Door Open', glyph:'/', passable:true,  art:'FLOOR'  },
  4:  { name:'Stairs ↓',  glyph:'>', passable:true,  art:'STAIRS' },
  5:  { name:'Stairs ↑',  glyph:'<', passable:true,  art:'STAIRS' },
  6:  { name:'Water',     glyph:'~', passable:false, art:'WATER'  },
  7:  { name:'Chest',     glyph:'$', passable:true,  art:'CHEST'  },
  8:  { name:'Altar',     glyph:'Ω', passable:true,  art:'SPECIAL'},
  9:  { name:'Pillar',    glyph:'O', passable:false, art:'PILLAR' },
  10: { name:'Void',      glyph:' ', passable:false, art:'VOID'   },
  11: { name:'Fire',      glyph:'¥', passable:false, art:'FIRE'   },
};

DCE.MapEntity = class {
  constructor(type, id, x, y, props={}) {
    this.entityType  = type;
    this.id          = id;
    this.x           = x;
    this.y           = y;
    this.properties  = props;
  }
};

DCE.TileMap = class {
  constructor(width, height, name='map') {
    this.width    = width;
    this.height   = height;
    this.name     = name;
    this.tiles    = Array.from({length:height}, ()=>Array(width).fill(DCE.TILE.FLOOR));
    this.revealed = Array.from({length:height}, ()=>Array(width).fill(false));
    this.entities = [];
    this.connections = {};
  }

  get(x, y) {
    if (x<0||y<0||x>=this.width||y>=this.height) return DCE.TILE.VOID;
    return this.tiles[y][x];
  }

  set(x, y, tile) {
    if (x<0||y<0||x>=this.width||y>=this.height) return;
    this.tiles[y][x] = tile;
  }

  isPassable(x, y) {
    if (x<0||y<0||x>=this.width||y>=this.height) return false;
    return DCE.TILE_DEFS[this.tiles[y][x]]?.passable ?? false;
  }

  revealRadius(cx, cy, radius=5) {
    for (let dy=-radius; dy<=radius; dy++) {
      for (let dx=-radius; dx<=radius; dx++) {
        if (dx*dx+dy*dy <= radius*radius) {
          const x=cx+dx, y=cy+dy;
          if (x>=0&&y>=0&&x<this.width&&y<this.height)
            this.revealed[y][x] = true;
        }
      }
    }
  }

  playerStart() {
    for (const e of this.entities)
      if (e.entityType==='player_start') return [e.x, e.y];
    // Find first passable tile
    for (let y=0; y<this.height; y++)
      for (let x=0; x<this.width; x++)
        if (this.isPassable(x,y)) return [x,y];
    return [1,1];
  }

  entitiesAt(x, y) {
    return this.entities.filter(e => e.x===x && e.y===y);
  }

  toDict() {
    return { name:this.name, width:this.width, height:this.height,
      tiles:this.tiles, revealed:this.revealed,
      entities:this.entities.map(e=>({type:e.entityType,id:e.id,x:e.x,y:e.y,props:e.properties})),
      connections:this.connections };
  }

  toJSON() { return JSON.stringify(this.toDict(), null, 2); }

  static fromDict(d) {
    const m = new DCE.TileMap(d.width, d.height, d.name||'map');
    m.tiles = d.tiles;
    m.revealed = d.revealed || Array.from({length:d.height},()=>Array(d.width).fill(false));
    m.entities = (d.entities||[]).map(e=>new DCE.MapEntity(e.type,e.id,e.x,e.y,e.props||{}));
    m.connections = d.connections||{};
    return m;
  }

  static fromJSON(json) { return DCE.TileMap.fromDict(JSON.parse(json)); }

  static blank(width, height, name='new_map') {
    const m = new DCE.TileMap(width, height, name);
    // Border walls
    for (let x=0; x<width; x++) { m.set(x,0,DCE.TILE.WALL); m.set(x,height-1,DCE.TILE.WALL); }
    for (let y=0; y<height; y++) { m.set(0,y,DCE.TILE.WALL); m.set(width-1,y,DCE.TILE.WALL); }
    return m;
  }

  static fromString(str, mapId='map') {
    const lines = str.trim().split('\n');
    const height = lines.length;
    const width  = Math.max(...lines.map(l=>l.length));
    const m = new DCE.TileMap(width, height, mapId);
    const charMap = {'#':DCE.TILE.WALL,'.':DCE.TILE.FLOOR,'+':DCE.TILE.DOOR,
      '>':DCE.TILE.STAIRS_D,'<':DCE.TILE.STAIRS_U,'~':DCE.TILE.WATER,
      '$':DCE.TILE.CHEST,'O':DCE.TILE.PILLAR,' ':DCE.TILE.FLOOR};
    for (let y=0; y<height; y++) {
      for (let x=0; x<lines[y].length; x++) {
        const ch = lines[y][x];
        if (ch === '@') {
          m.set(x,y,DCE.TILE.FLOOR);
          m.entities.push(new DCE.MapEntity('player_start','player_start',x,y));
        } else {
          m.set(x, y, charMap[ch] ?? DCE.TILE.FLOOR);
        }
      }
    }
    return m;
  }
};
